sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,343 @@
+        {
+            var type = parameterSymbol.Type;
+            var isExcluded =
+                type.DerivesFrom(KnownType.System_Array) ||
+                type.IsValueType ||
+                type.Is(KnownType.System_String);
+
+            return !isExcluded;
+        }
+
+        private ITypeSymbol FindParameterUseAsType(IdentifierNameSyntax idRef, SemanticModel semanticModel)
+        {
+            if (idRef.Parent is ConditionalAccessExpressionSyntax)
+            {
+                var cond = idRef.Parent as ConditionalAccessExpressionSyntax;
+                var binding = cond.WhenNotNull as MemberBindingExpressionSyntax;
+
+                if (binding != null)
+                {
+                    var name = binding.Name;
+                    var accessedMember = semanticModel.GetSymbolInfo(name).Symbol;
+
+                    var propertySymbol = accessedMember as IPropertySymbol;
+                    if (propertySymbol != null)
+                    {
+                        return HandleProperty(idRef, propertySymbol);
+                    }
+
+                    return FindOriginatingSymbol(accessedMember);
+                }
+                else if (cond.WhenNotNull is InvocationExpressionSyntax)
+                {
+                    var inv = (cond.WhenNotNull as InvocationExpressionSyntax).Expression as MemberBindingExpressionSyntax;
+
+                    if (inv != null)
+                    {
+                        var invocationSymbol = semanticModel.GetSymbolInfo(inv).Symbol;
+                        return HandleInvocation(idRef, invocationSymbol, semanticModel);
+                    }
+                }
+            }
+            else if (idRef.Parent is MemberAccessExpressionSyntax)
+            {
+                var invocation = idRef.Parent.Parent as InvocationExpressionSyntax;
+                if (invocation != null)
+                {
+                    var invocationSymbol = semanticModel.GetSymbolInfo(idRef.Parent.Parent).Symbol;
+                    return HandleInvocation(idRef, invocationSymbol, semanticModel);
+                }
+
+                var accessedMember = semanticModel.GetSymbolInfo(idRef.Parent).Symbol;
+
+                var propertySymbol = accessedMember as IPropertySymbol;
+                if (propertySymbol != null)
+                {
+                    return HandleProperty(idRef, propertySymbol);
+                }
+
+                return FindOriginatingSymbol(accessedMember);
+            }
+            else if (idRef.Parent is ArgumentSyntax)
+            {
+                return semanticModel.GetTypeInfo(idRef).ConvertedType;
+            }
+
+            return null;
+        }
+
+        private ITypeSymbol HandleProperty(IdentifierNameSyntax idRef, IPropertySymbol propertySymbol)
+        {
+            return FindOriginatingSymbol(idRef.Parent.Parent is AssignmentExpressionSyntax
```

Comment:
Same here about `idRef.Parent.Parent`, are you sure not to have any null in the path? Are you sure you don't have an issue when there are parenthesis?